### DISCUSSION TITLE
Tmux-like Wezterm Multiplexing Sessions

### DISCUSSION THREAD

**Discussion by kitten** [wezterm/wezterm#1322] [Show and tell]
I worked on a little implementation of Tmux-like sessions for Wezterm via a quick and dirty shell script, and I'd love to show it off here for multiple reasons, but mainly because I think it could start some discussions and there was some interest in this already.

Please do chime in to tell me there's an easier way to implement this, if there is 😆 I've gone off of the existing docs and information from issues and discussion threads, but I haven't read the `wezterm` codebase extensively, so I'm just learning more about it by doing this.

Mainly, I've seen that two prior discussions relate to this topic:
- https://github.com/wez/wezterm/discussions/1017
- https://github.com/wez/wezterm/discussions/712
- https://github.com/wez/wezterm/discussions/1318

How my script works is pretty similar to a tmux-like workflow:
- New `wezterm-mux-server`s must be spawnable as daemons
- It must be possible to close/end/kill sessions selectively when a command is invoked
- The session must die when the `wezterm-mux-server` instance dies
- I must be able to list out all open sessions
- I must be able to attach to running sessions

To just get to the point, **here is the script in question:** https://github.com/kitten/nix-system/blob/abe3322f8991fe54e681f091d7e3bd438fcb60df/config/scripts/wezmux.sh

### How does it work?

This script "exploits" the existing `unix_domains` config. It does not integrate with Wezterm itself _that well_, and instead runs `wezterm` or `wezterm-mux-server` manually. I'm on macOS, hence I haven't spent too much time on launching new Wezterm GUIs on other platforms, since on macOS I can simply call `open -a ... --args`, which isn't perfect but works for now.

All sessions are kept in `.local/share/wezterm/sessions/*`. Each session has a folder that contains a `pid` file with the daemon's PID, and a logfile. Creating the `sock` socket file is a responsibility of `wezterm-mux-server` itself. I don't let the server daemonize itself, but instead, the bash script itself becomes a small "faux daemon" and puts its own PID into the `pid` file.

Because of the folder structure, I can list out running sessions. Each time a new session is created, the bash script spawns a daemonized version of itself that traps exit signals to a small callback function that cleans up the session folder, and terminates the `wezterm-mux-server` as needed. The `wexterm-mux-server` is started with an overriden `--config "unix_domains=..."` config entry, which points at this new `sessions/*/` folder.

The main script can then attach to the sessions by launching the GUI executable directly, on macOS via: `open -a $WEZTERM_EXECUTABLE -n --args ...`.

What I've discovered is that this workflow is actually _extremely comparable_ to tmux. I can launch a Wezterm window without a mux server (which I call default) and use it to create new sessions and attach to those. Once the session's Wezterm GUI window is spawned, I can exit the default (sessionless) window manually. The session stays active even when I close this new window and I can repeatedly attach to it.

### How well does it integrate?

There's a bunch of things I do to my `wezterm.lua` configuration to make this a little easier to handle.
I had to, for instance, create a small status_right modification that reads `unix_domains` to display the currently active session's name: https://github.com/kitten/nix-system/blob/abe3322f8991fe54e681f091d7e3bd438fcb60df/config/home/wezterm/init.lua#L5-L13

Apart from that there's several things that I wish I could do:
- Wezterm has an open GUI socket for the active window, but I can't open a window in the currently running instance of wezterm and pass it the unix domain to connect to
- Wezterm has a launcher menu, but I believe it'd be a little hacky at best to update its content (if that's possible; there's no event for it, so it'd have to override the window's config periodically?) to add entries for all running sessions
- Wezterm can't detach/reattach the current GUI to and from sessions when it's connected to domains (or kill the active window programmatically). This isn't that big of a deal and an addition of opening new windows for sockets would maybe make this obsolete 🤔 

So, I'm personally pretty happy with this admittedly super hacky solution 😄, but I think this could hopefully help form a better idea of what capabilities the GUI is lacking, in terms of its API and the GUI socket.

**Comment by wez** [comment]
Thanks for putting this together!

I'm going to digest this a bit.  I'd love to hear what thoughts you have on making this sort of functionality a first-class thing in wezterm; especially the parts that were awkward or that don't work quite how you'd like them to in your mux script!

**Reply by kitten** [reply]
I think it'd be awesome to get this working first class! What I'm unclear on is what `wezterm-mux-server` exactly does with its configuration; If I'm not mistaken it will simply fork and initialise for all available `unix_domains`? https://github.com/wez/wezterm/blob/e7b8c7e8b89252cf88fd9719724277abd6d9d973/wezterm-mux-server/src/main.rs#L215-L221

I suppose the question then becomes — given the script above that I've posted — if Wezterm was to make a minimum change, it could simply allow `--daemonize` to accept a target `socket_path` and `name` to spin up a "dynamic" unix domain. The problem is that that may influence how `daemon_options` must be used so PID files don't override eachother.

So there's several choices here to be made; if I'm not misunderstanding `wezterm-mux-server`, and it does spin up multiple sockets per item in `unix_domains`, then the problem becomes that `daemon_options` would also be dependent on that, or that `wezterm-mux-server --daemonize` would have to accept many overrides still
(I did try `--daemonize` before, but couldn't quite get the overrides for `daemon_options` to work or get it to connect to create the socket; if that works already, a more "complete" CLI addition could already do what the script does right now)

There's also the possibility of Wezterm becoming more opinionated about unix domains, which would allow it to create a folder structure of its own for multiple unix domains in the same environment dynamically. But I simply don't know enough about the extent of the Wezterm Multiplexing use-cases to know whether that's even an option 😅

**Reply by wez** [reply]
In the current nightly builds, wezterm now has the concept of a *workspace*.  A workspace is a collection of Windows (and their Tabs and Panes).

You can use `wezterm.action{SwitchToWorkspace={name="coding"}}` to switch to a workspace named `coding`, creating it if it doesn't exist.

`wezterm.action{SwitchToWorkspace={}}` will create a workspace with a generated name and switch to it.

Newly created workspaces will spawn the default program.  You can use `wezterm.action{SwitchToWorkspace={spawn={args={'top'}}}` to create a new workspace that runs `top` for example.  The `spawn` field is any valid [SpawnCommand](https://wezfurlong.org/wezterm/config/lua/SpawnCommand.html).

The launcher menu (right click on the new tab button, or otherwise assign a key to [ShowLauncher](https://wezfurlong.org/wezterm/config/lua/keyassignment/ShowLauncher.html)) has items to activate existing workspaces and to create a new workspace.

The active workspace acts as a filter on the windows that are shown in the gui.  Only windows that are part of the active workspace are shown.

When the active workspace is empty, the gui will select and activate some other workspace rather than exit.  (there's probably some work to do around this: feedback is welcome!)

Workspaces work out of the box with no configuration.  They also work with multiplexing domains with no additional configuration beyond setting up the multiplexer domain as normal.

I've love to have your feedback on this!

https://github.com/wez/wezterm/issues/1531 is the issue where I'm tracking this feature set.

**Reply by gegoune** [reply]
Is there, or is it planned, a cli interface so one can use for example fzf to navigate between workspaces?

**Reply by gegoune** [reply]
And another question. Is default workspace being created on Wezterm's startup? If so, how can I switch to it with `SwitchToWorkspace`?

**Reply by wez** [reply]
> Is there, or is it planned, a cli interface so one can use for example fzf to navigate between workspaces?

The launcher menu lists workspaces.  In the launcher menu you can press `/` to enter fuzzy matching mode.
I plan to add a new key assignment mode that will let you run the launcher and list only workspaces.

**Reply by wez** [reply]
> And another question. Is default workspace being created on Wezterm's startup? If so, how can I switch to it with `SwitchToWorkspace`?

Yes; the default workspace is named `default`.  You can use `wezterm.action{SwitchToWorkspace={name="default"}}` to switch to it.

You can use `wezterm start --workspace NAME` to change that name to something else.  Similarly, `wezterm cli spawn --workspace NAME` can be used to do that via the CLI.

**Reply by gegoune** [reply]
Would be nice to have an option to provide name for default workspace. I only ever start wezterm from macOS UI (not from another terminal).

Action to list existing workspaces would be nice, will use launcher for now. Thank you! that's one great addition!

**Reply by wez** [reply]
https://github.com/wez/wezterm/commit/7eec41c284a47721ed51f5305b7106a59dae2f84 allows for this, which shows the launcher in fuzzy matching mode, scoped only to workspaces:

```lua
local wezterm = require 'wezterm'

return {
  keys = {
    {key="9", mods="ALT", action=wezterm.action{ShowLauncherArgs={flags="FUZZY|WORKSPACES"}}},
  },
}
```

**Reply by gegoune** [reply]
Sorry to bombard you with those, but just thought that it would be really nice to be able to add current workflow's name to statusbar (classic one for me :)).

**Reply by gegoune** [reply]
And maybe one more. How about ordering (automatically) created workspace, so one can switch between them by some next/prev mapping?

**Reply by wez** [reply]
https://wezfurlong.org/wezterm/config/lua/window/active_workspace.html
https://wezfurlong.org/wezterm/config/lua/config/default_workspace.html
https://wezfurlong.org/wezterm/config/lua/keyassignment/SwitchWorkspaceRelative.html

**Reply by gegoune** [reply]
Amazing, thank you so much! Sort of unrelated question, there isn't `set_left_status`, is there?

**Reply by wez** [reply]
The status area only exists to the right of tabs today.  I don't have any plans to add other kinds of status areas.

**Reply by gegoune** [reply]
Perhaps I could tempt you a bit with approach such as status bar being fully configurable and tab being returned by some function, and instead of having `set_right_status` one would use some delimiter character. Something similar to vim's status line maybe?

Reason why I am asking about that is because I would like to have menubar such as:

`[workspacename] [tabs]    >(right hand)[time/date]`

**Reply by wez** [reply]
I'm not keen to parse things out of text like that, but I'm open to the idea of adding an official left status bar; please file an enhancement issue for that.

**Reply by gegoune** [reply]
Is there any way to interactively create new workspace with key mapping (with being asked for workspace name)?

**Reply by wez** [reply]
Not directly in the UI.  The closest  you could get today is `wezterm cli spawn --new-window --workspace NAME`; that will create a window in the workspace but not activate it.

**Comment by joexue** [comment]
Let's have the "tmux -CC " please, then all are resolved.

**Reply by cazador481** [reply]
An ability to rename an existing workspace would be nice.

**Reply by zyriab** [reply]
You can do it now with `wezterm cli rename-workspace your_name`

**Comment by michaelbrusegard** [comment]
Would it be possible to setup workspaces to be tied to a single window (and not hide other windows) in a similar manner to how it is done in tmux sessions. I tried using workspaces, but they can be annoying when using a tiling window manager. If I have a wezterm window open on another desktop it will also be hidden when I switch workspace., and when I reopen the workspace all the windows become a mess because of the tiling window manager. I really like using wezterm so far, and having sessions would be lovely
